File: programming/project-learning/perl/algorithms/priority-queue.pl
package PriorityQueue;

use strict;
use warnings;
use Carp;

sub new {
    my ($class, $compare) = @_;
    croak "Compare has to be a function" unless ref($compare) eq 'CODE';
    my $self = {
        heap => [],
        compare => $compare,
    };
    bless $self, $class;
    return $self;
}

sub get_parent_index {
    my ($self, $index) = @_;
    return int(($index - 1) / 2);
}

sub get_left_child_index {
    my ($self, $index) = @_;
    return 2 * $index + 1;
}

sub get_right_child_index {
    my ($self, $index) = @_;
    return 2 * $index + 2;
}

sub swap {
    my ($self, $index1, $index2) = @_;
    ($self->{heap}[$index1], $self->{heap}[$index2]) = ($self->{heap}[$index2], $self->{heap}[$index1]);
}

1;

File: programming/project-learning/c++/algorithms/priority-queue.cpp
#include <vector>
#include <functional>
#include <stdexcept>

class PriorityQueue {
public:
    PriorityQueue(std::function<bool(int, int)> compare) : compare(compare) {
        if (!compare) {
            throw std::invalid_argument("Compare has to be a function");
        }
    }

    int getParentIndex(int index) {
        return (index - 1) / 2;
    }

    int getLeftChildIndex(int index) {
        return 2 * index + 1;
    }

    int getRightChildIndex(int index) {
        return 2 * index + 2;
    }

    void swap(int index1, int index2) {
        std::swap(heap[index1], heap[index2]);
    }

private:
    std::vector<int> heap;
    std::function<bool(int, int)> compare;
};